compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs

```diff
@@ -1955,6 +1964,410 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+        let relevant_impl_generics = impl_generics.params.iter();
+
+        let relevant_impl_generics: Vec<&ty::GenericParamDef> = match relevant_broken_predicate {
+            ty::PredicateKind::Clause(ty::Clause::Trait(broken_trait)) => relevant_impl_generics
+                .filter(|&generic| {
+                    // Only retain generics that are mentioned in the (Self type) of this predicate:
+                    find_param_def_in_ty_walker(broken_trait.trait_ref.self_ty().walk(), generic)
+                })
+                .collect(),
+            _ => {
+                // Because we are missing information about which might be relevant, we assume that they all are.
+                relevant_impl_generics.collect()
+            }
+        };
+
+        enum PointableType<'tcx> {
+            Adt(ty::AdtDef<'tcx>),
+            Tuple,
+        }
+
+        // Note that there is no point in supporting "primitive" types like char/i32,
+        // since we cannot refine such a span any more anyway.
+        // Right now, only ADTs (struct/enum (variant) constructors) and tuples are supported by this refinement.
+        let (impl_self_ty, impl_self_ty_args) = match impl_self_ty.kind() {
+            ty::Adt(impl_self_ty_path, impl_self_ty_args) => (
+                PointableType::Adt(*impl_self_ty_path),
+                impl_self_ty_args.iter().collect::<Vec<_>>(),
+            ),
+            ty::Tuple(impl_self_ty_args) => (
+                PointableType::Tuple,
+                impl_self_ty_args.iter().map(|t| ty::GenericArg::from(t)).collect::<Vec<_>>(),
+            ),
+            _ => {
+                return Err(expr);
+            }
+        };
+
+        // We need to find which of the type's arguments are relevant to this obligation.
+        // For example, if we had an impl for `SomeTime<(A, C), B, Result<B, C>>` and the broken obligation
+        // was `B: Display` then we'd care about indexes `vec![1, 2]`, but if it was `C: PartialEq` then we'd
+        // care about index `vec![0, 2]`.
+        let relevant_ty_args_indices: Vec<usize> = impl_self_ty_args
+            .into_iter()
+            .enumerate()
+            .filter(|(_index, ty_arg)| {
+                relevant_impl_generics
+                    .iter()
+                    .any(|param| find_param_def_in_ty_walker(ty_arg.walk(), param))
+            })
+            .map(|(index, _ty_arg)| index)
+            .collect();
+
+        // Different expression require different treatment. In general, we hope that we have a literal
+        // expression which produces the current `Self` type. Moreoever, we hope that it has a single
+        // field which can be "blamed" for the missing trait, so that it can be highlighted.
+        match expr.kind {
+            hir::ExprKind::Struct(
+                hir::QPath::Resolved(
+                    _,
+                    hir::Path { res: hir::def::Res::Def(struct_def_kind, struct_def_id), .. },
```

Comment:
There should be a way of extracting the def here rather than destructuring the path's res -- try [`type_dependent_def`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/typeck_results/struct.TypeckResults.html#method.type_dependent_def)

Reply:
I used `.qpath_res` here instead; `type_dependent_def` doesn't seem to work, I think because this isn't a call?

There's a second use of pretty much the same logic below, where `type_dependent_def` might work. I haven't tried that yet.

Reply:
My bad -- `qpath_res` is the one I was thinking of, yeah.